### PR TITLE
nixos/pam: enable unlocking ZFS home dataset

### PR DESCRIPTION
**Pull request by tensor5**
###### Description of changes

This enables automatic unlocking and mounting at login of a ZFS home dataset encrypted with user's password.

Inspired by https://github.com/benpye/nixpkgs/tree/add-zfs-home-mount.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Please add a NixOS test, PAM modifications are quite dangerous, and we had breakages in the past due to them. I don't have enough time to review this manually in a reasonable timeframe, a NixOS test would make my review much easier (and the subsequent contribution process too).

File: nixos/modules/security/pam.nix [hunk 1]
```diff
@@ -1202,6 +1222,34 @@ in
...
+          Enable unlocking and mounting of ZFS home dataset at login.
+        '';
+      };
+
+      homes = mkOption {
+        example = "rpool/home";
+        default = "rpool/home";
+        type = types.str;
+        description = lib.mdDoc ''
+          Prefix of home datasets. This value will be contacenated with
```
Comment (RaitoBezarius) [inline nixos/modules/security/pam.nix:1239]:
```suggestion
          Prefix of home datasets. This value will be concatenated with
```

File: nixos/modules/security/pam.nix [hunk 2]
```diff
@@ -1202,6 +1222,34 @@ in
       };
     };
 
+    security.pam.zfs = {
```
Comment (RaitoBezarius) [inline nixos/modules/security/pam.nix:1225]: There should be an assertion somewhere to detect if the user has indeed ZFS supported on the system, it does not make sense to have those modules loaded without ZFS support.

File: pkgs/os-specific/linux/zfs/default.nix [hunk 3]
```diff
@@ -137,6 +138,7 @@ let
         "--sysconfdir=/etc"
         "--localstatedir=/var"
         "--enable-systemd"
+        "--enable-pam"
```
Comment (RaitoBezarius) [inline pkgs/os-specific/linux/zfs/default.nix:141]: What is the closure size increase by adding PAM to the build?

Comment (tensor5) [reply pkgs/os-specific/linux/zfs/default.nix:141]: The effect of `--enable-pam` is adding `lib/security/pam_zfs_key.so` which is about 73 KB.

File: nixos/modules/security/pam.nix (hunk 2 above)
Comment (tensor5) [reply nixos/modules/security/pam.nix:1225]: The option is of by default, anyways.

File: nixos/modules/security/pam.nix (hunk 1 above)
Comment (tensor5) [reply nixos/modules/security/pam.nix]: Fixed in 5cdfd758d576cceacaee94d290c2e2a6a7d1ee7b.

File: nixos/modules/security/pam.nix (hunk 2 above)
Comment (tensor5) [reply nixos/modules/security/pam.nix:1225]: a4962a529ed3ba577c80754acc2746b0b8764c81

File: pkgs/os-specific/linux/zfs/default.nix (hunk 3 above)
Comment (RaitoBezarius) [reply pkgs/os-specific/linux/zfs/default.nix:141]: Awesome

**Review by RaitoBezarius — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 228956` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>libvmi</li>
    <li>nixops_unstable</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>71 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>booster</li>
    <li>ceph (ceph-dev)</li>
    <li>ceph-client</li>
    <li>ceph-csi</li>
    <li>ceph.dev (ceph-dev.dev)</li>
    <li>ceph.doc (ceph-dev.doc)</li>
    <li>libceph (ceph.lib ,libceph.dev ,libceph.doc ,libceph.lib ,libceph.man)</li>
    <li>ceph.man (ceph-dev.man)</li>
    <li>check_zfs</li>
    <li>collectd</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>docker-machine-kvm</li>
    <li>docker-machine-kvm2</li>
    <li>easysnap</li>
    <li>gnome.gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>haskellPackages.libvirt-hs</li>
    <li>haskellPackages.libvirt-hs.doc</li>
    <li>libguestfs</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>minishift</li>
    <li>multipass</li>
    <li>partition-manager</li>
    <li>perl534Packages.SysVirt</li>
    <li>perl534Packages.SysVirt.devdoc</li>
    <li>perl536Packages.SysVirt</li>
    <li>perl536Packages.SysVirt.devdoc</li>
    <li>python310Packages.guestfs</li>
    <li>python310Packages.guestfs.dist</li>
    <li>python310Packages.libvirt</li>
    <li>python310Packages.libvirt.dist</li>
    <li>python310Packages.py-libzfs</li>
    <li>python310Packages.py-libzfs.dist</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.libvirt</li>
    <li>python311Packages.libvirt.dist</li>
    <li>python311Packages.py-libzfs</li>
    <li>python311Packages.py-libzfs.dist</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>rubyPackages.ruby-libvirt (rubyPackages_3_1.ruby-libvirt)</li>
    <li>rubyPackages_2_7.ruby-libvirt</li>
    <li>rubyPackages_3_0.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>sambaFull (samba4Full)</li>
    <li>sambaFull.dev (samba4Full.dev)</li>
    <li>sambaFull.man (samba4Full.man)</li>
    <li>sanoid</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-manager.dist</li>
    <li>virt-top</li>
    <li>virt-viewer</li>
    <li>zfs (zfsStable)</li>
    <li>zfs.dev (zfsStable.dev)</li>
    <li>zfsUnstable</li>
    <li>zfsUnstable.dev</li>
    <li>zfstools</li>
    <li>zpool-auto-expand-partitions</li>
    <li>zxfer</li>
  </ul>
</details>

**Review by RaitoBezarius — APPROVED** [review]

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Still no NixOS test though.

**Comment by RaitoBezarius** [comment]
@tensor5 Could you cleanup the conflicts please?

**Comment by tensor5** [comment]
@RaitoBezarius, do you prefer a squashed commit?

**Comment by tensor5** [comment]
Done.

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 228956` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>71 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>booster</li>
    <li>ceph</li>
    <li>ceph-client</li>
    <li>ceph-csi</li>
    <li>ceph.dev</li>
    <li>ceph.doc</li>
    <li>libceph (ceph.lib ,libceph.dev ,libceph.doc ,libceph.lib ,libceph.man)</li>
    <li>ceph.man</li>
    <li>check_zfs</li>
    <li>collectd</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>docker-machine-kvm</li>
    <li>docker-machine-kvm2</li>
    <li>easysnap</li>
    <li>gnome.gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>minishift</li>
    <li>multipass</li>
    <li>ocamlPackages.ocaml_libvirt</li>
    <li>partition-manager</li>
    <li>perl534Packages.SysVirt</li>
    <li>perl534Packages.SysVirt.devdoc</li>
    <li>perl536Packages.SysVirt</li>
    <li>perl536Packages.SysVirt.devdoc</li>
    <li>python310Packages.guestfs</li>
    <li>python310Packages.guestfs.dist</li>
    <li>python310Packages.libvirt</li>
    <li>python310Packages.libvirt.dist</li>
    <li>python310Packages.py-libzfs</li>
    <li>python310Packages.py-libzfs.dist</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.libvirt</li>
    <li>python311Packages.libvirt.dist</li>
    <li>python311Packages.py-libzfs</li>
    <li>python311Packages.py-libzfs.dist</li>
    <li>qemu_full</li>
    <li>qemu_full.debug</li>
    <li>qemu_full.ga</li>
    <li>rubyPackages.ruby-libvirt</li>
    <li>rubyPackages_2_7.ruby-libvirt</li>
    <li>rubyPackages_3_0.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>samba4Full</li>
    <li>samba4Full.dev</li>
    <li>samba4Full.man</li>
    <li>sanoid</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-manager.dist</li>
    <li>virt-top</li>
    <li>virt-viewer</li>
    <li>zfs</li>
    <li>zfs.dev</li>
    <li>zfsUnstable</li>
    <li>zfsUnstable.dev</li>
    <li>zfstools</li>
    <li>zpool-auto-expand-partitions</li>
    <li>zxfer</li>
  </ul>
</details>

**Pull request merged**